In [79]:
import pandas as pd
import numpy as np

Load Dataset

In [80]:
df = pd.read_csv("../data/CineStream_Catalog.csv")

In [81]:
df.head()

,ContentID,Title,Type,Genre,Language,Country,Director,CastCount,ReleaseYear,AddedDate,...,IMDbScore,ViewsMillions,WatchHoursMillions,AvgCompletionPct,AwardsWon,ProductionCostCr,RevenueCr,SubscribersGainedThousands,TrendingStatus,CriticRating
0,CS-0001,Voyage Whispering II,Movie,Kids,Hindi,India,Akira Brown,3,2024,2020-04-28,...,8.0,25.28,40.81,65,4,56.1,59.5,5873,Classic,7.4
1,CS-0002,Endless River: Reborn,Series,Kids,Korean,South Korea,Riya Patel,6,2017,2020-07-30,...,6.8,24.13,244.20,69,0,43.6,33.6,-1000,New,6.2
2,CS-0003,Distant Letter of the North,Series,Romance,Telugu,India,Liam Iyer,4,2016,2020-08-29,...,7.7,26.77,107.08,80,2,63.6,56.6,1016,Popular,6.9
3,CS-0004,Light Whispering,Series,Animation,English,Australia,Min-jun Khan,5,2020,2020-02-16,...,5.6,8.67,111.27,55,0,44.3,8.6,-1000,Trending,6.6
4,CS-0005,Game Quiet III,Stand-up,Comedy,Hindi,India,Yuki Yilmaz,13,2017,2021-04-20,...,5.5,7.01,5.29,52,0,1.7,10.6,1976,Classic,4.8


Dataset Shape Check

In [82]:
print("Rows and Columns:")
print(df.shape)

Rows and Columns:
(524, 23)


Column Information

In [83]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 524 entries, 0 to 523
Data columns (total 23 columns):
 #   Column                      Non-Null Count  Dtype  
---  ------                      --------------  -----  
 0   ContentID                   524 non-null    object 
 1   Title                       524 non-null    object 
 2   Type                        524 non-null    object 
 3   Genre                       524 non-null    object 
 4   Language                    524 non-null    object 
 5   Country                     524 non-null    object 
 6   Director                    520 non-null    object 
 7   CastCount                   524 non-null    int64  
 8   ReleaseYear                 524 non-null    int64  
 9   AddedDate                   521 non-null    object 
 10  RuntimeMinutes              524 non-null    int64  
 11  EpisodeCount                524 non-null    int64  
 12  AgeRating                   524 non-null    object 
 13  IMDbScore                   524 non

In [84]:
df.describe()

,CastCount,ReleaseYear,RuntimeMinutes,EpisodeCount,IMDbScore,ViewsMillions,WatchHoursMillions,AvgCompletionPct,AwardsWon,ProductionCostCr,RevenueCr,SubscribersGainedThousands,CriticRating
count,524.000000,524.000000,524.000000,524.000000,524.000000,524.000000,524.000000,524.000000,524.000000,524.000000,524.000000,524.000000,524.000000
mean,8.761450,2019.465649,481.116412,6.047710,7.091603,16.106183,47.353473,60.303435,1.547710,39.144275,25.017939,3126.242366,6.795038
std,3.801776,2.856806,4371.900126,9.650866,1.131204,9.855031,73.472515,12.973628,1.566748,26.307340,17.876782,4339.804813,1.484984
min,3.000000,2015.000000,50.000000,1.000000,3.700000,0.200000,0.100000,15.000000,0.000000,0.500000,0.100000,-50000.000000,1.700000
25%,5.000000,2017.000000,106.750000,1.000000,6.300000,8.317500,10.692500,52.000000,0.000000,19.175000,11.350000,1126.000000,5.800000
50%,9.000000,2019.500000,149.000000,1.000000,7.100000,15.800000,23.025000,60.000000,1.000000,37.050000,21.850000,3241.000000,6.700000
75%,12.000000,2022.000000,300.000000,8.000000,7.800000,22.920000,48.557500,69.250000,3.000000,53.650000,34.550000,5270.500000,7.800000
max,15.000000,2024.000000,99999.000000,40.000000,11.400000,56.630000,633.910000,98.000000,9.000000,144.100000,82.300000,11636.000000,12.900000


In [85]:
df.columns

Index(['ContentID', 'Title', 'Type', 'Genre', 'Language', 'Country',
       'Director', 'CastCount', 'ReleaseYear', 'AddedDate', 'RuntimeMinutes',
       'EpisodeCount', 'AgeRating', 'IMDbScore', 'ViewsMillions',
       'WatchHoursMillions', 'AvgCompletionPct', 'AwardsWon',
       'ProductionCostCr', 'RevenueCr', 'SubscribersGainedThousands',
       'TrendingStatus', 'CriticRating'],
      dtype='object')

In [86]:
df.dtypes

ContentID                      object
Title                          object
Type                           object
Genre                          object
Language                       object
Country                        object
Director                       object
CastCount                       int64
ReleaseYear                     int64
AddedDate                      object
RuntimeMinutes                  int64
EpisodeCount                    int64
AgeRating                      object
IMDbScore                     float64
ViewsMillions                 float64
WatchHoursMillions            float64
AvgCompletionPct                int64
AwardsWon                       int64
ProductionCostCr              float64
RevenueCr                     float64
SubscribersGainedThousands      int64
TrendingStatus                 object
CriticRating                  float64
dtype: object

Missing Values

In [87]:
df.isnull().sum()

ContentID                     0
Title                         0
Type                          0
Genre                         0
Language                      0
Country                       0
Director                      4
CastCount                     0
ReleaseYear                   0
AddedDate                     3
RuntimeMinutes                0
EpisodeCount                  0
AgeRating                     0
IMDbScore                     0
ViewsMillions                 0
WatchHoursMillions            0
AvgCompletionPct              0
AwardsWon                     0
ProductionCostCr              0
RevenueCr                     0
SubscribersGainedThousands    0
TrendingStatus                0
CriticRating                  0
dtype: int64

Duplicate Rows

In [88]:
df.duplicated().sum()

np.int64(4)

check categorical columns unique values count

In [89]:
df[['Type','Genre','Language','Country','AgeRating','TrendingStatus']].value_counts()

Type         Genre      Language  Country      AgeRating  TrendingStatus
Movie        Drama      Korean    South Korea  U/A 13+    Classic           4
             Thriller   English   UK           U/A 16+    Popular           3
             Animation  Hindi     India        U/A 13+    New               2
                                                          Popular           2
             Comedy     Hindi     India        U/A 13+    Classic           2
                                                                           ..
Documentary  Comedy     Telugu    India        U/A 13+    New               1
                                               U/A 7+     Classic           1
             Comedy     Tamil     India        U/A 13+    Popular           1
             Drama      Japanese  Japan        U/A 16+    Popular           1
series       Thriller   Kannada   India        U/A 7+     New               1
Name: count, Length: 482, dtype: int64

Unique Values of Categorical Columns

In [90]:
categorical_cols = ['Type','Genre','Language','Country','AgeRating','TrendingStatus']


In [91]:
for col in categorical_cols:
    print("\n", col)
    print(df[col].unique())


 Type
['Movie' 'Series' 'Stand-up' 'movie' 'Documentary' 'series' 'documentary']

 Genre
['Kids' 'Romance' 'Animation' 'Comedy' 'Action' 'Sci-Fi' 'Thriller'
 'Reality' 'Drama' 'Action ' 'Horror' 'Romance ' 'Drama ' 'Comedy '
 'Reality ' 'Horror ' 'Thriller ' 'Kids ' 'Sci-Fi ']

 Language
['Hindi' 'Korean' 'Telugu' 'English' 'Tamil' 'Turkish' 'Bengali'
 'Malayalam' 'Hindi ' 'Kannada' 'Spanish' 'Japanese' 'Telugu ' 'Tamil '
 'English ' 'Turkish ' 'Kannada ']

 Country
['India' 'South Korea' 'Australia' 'Turkey' 'UK' 'USA' 'Spain' 'Canada'
 'Japan']

 AgeRating
['U/A 13+' 'U/A 7+' 'U/A 16+' 'U' 'A']

 TrendingStatus
['Classic' 'New' 'Popular' 'Trending']


issues:

In [92]:
print("Invalid IMDb Scores:",
      ((df['IMDbScore'] < 1) | (df['IMDbScore'] > 10)).sum())

print("Invalid Critic Ratings:",
      ((df['CriticRating'] < 1) | (df['CriticRating'] > 10)).sum())

print("Negative Subscribers:",
      (df['SubscribersGainedThousands'] < 0).sum())

print("\nTop Runtime Values:")
print(df['RuntimeMinutes'].sort_values(ascending=False).head())

Invalid IMDb Scores: 3
Invalid Critic Ratings: 2
Negative Subscribers: 81

Top Runtime Values:
421    99999
425     2200
373     2200
349     2200
150     2200
Name: RuntimeMinutes, dtype: int64


#### Data Quality Report

Issues Found:

1. Type column contains case mismatches.
2. Genre column contains trailing spaces.
3. Language column contains trailing spaces.
4. Director column contains missing values.
5. AddedDate column contains missing values.
6. IMDbScore contains 3 values outside the valid range (1–10).
7. CriticRating contains 2 values outside the valid range (1–10).
8. RuntimeMinutes contains an extreme outlier value (99999 minutes).
9. SubscribersGainedThousands contains 81 negative values.
10. Duplicate rows found.

Start Cleaning:

In [93]:
clean_df = df.copy()

strip whitespace in text columns

In [94]:
text_cols = ['Title',
    'Type',
    'Genre',
    'Language',
    'Country',
    'Director',
    'AgeRating',
    'TrendingStatus'
]

In [95]:
for col in text_cols:
    clean_df[col] = clean_df[col].str.strip()

Standardise case in Type Column

In [96]:
clean_df['Type'] = clean_df['Type'].str.title()

Drop duplicates

In [97]:
clean_df = clean_df.drop_duplicates()

In [98]:
clean_df.duplicated().sum()

np.int64(0)

Replace out-of-range ratings and runtimes with NaN

In [99]:
clean_df.loc[
    (clean_df['IMDbScore'] < 1) |
    (clean_df['IMDbScore'] > 10),
    'IMDbScore'
] = np.nan

In [100]:
clean_df.loc[
    (clean_df['CriticRating'] < 1) |
    (clean_df['CriticRating'] > 10),
    'CriticRating'
] = np.nan

In [101]:
clean_df.loc[
    clean_df['RuntimeMinutes'] > 5000,
    'RuntimeMinutes'
] = np.nan

In [102]:
clean_df['RuntimeMinutes'][clean_df['RuntimeMinutes'].isna()]

421   NaN
Name: RuntimeMinutes, dtype: float64

Replace negative subscriber counts with 0

In [103]:
clean_df.loc[
    clean_df['SubscribersGainedThousands'] < 0,
    'SubscribersGainedThousands'
] = 0

Convert AddedDate to a real datetime.

In [105]:
clean_df['AddedDate'] = pd.to_datetime(clean_df['AddedDate'],
    errors='coerce'
)

Create Profit_Cr Column

In [106]:
clean_df['Profit_Cr'] = (clean_df['RevenueCr'] - clean_df['ProductionCostCr'])

Create ROI_Pct Column

In [107]:
clean_df['ROI_Pct'] = (clean_df['Profit_Cr'] / clean_df['ProductionCostCr']
) * 100

Create Performance_Band Column

In [108]:
clean_df['Performance_Band'] = np.where(clean_df['Profit_Cr'] > 20,'Hit',
    np.where(clean_df['Profit_Cr'] >= 0,'Break-even','Flop')
)

Check New Columns

In [109]:
clean_df[['Profit_Cr','ROI_Pct','Performance_Band']].head()

,Profit_Cr,ROI_Pct,Performance_Band
0,3.4,6.060606,Break-even
1,-10.0,-22.935780,Flop
2,-7.0,-11.006289,Flop
3,-35.7,-80.586907,Flop
4,8.9,523.529412,Break-even


Save Clean Dataset

In [110]:
clean_df.to_csv('../outputs/cleaned_cinestream.csv',index=False)

Final Verification

In [114]:
print(clean_df.shape)

(520, 26)


In [115]:
clean_df.info()

<class 'pandas.core.frame.DataFrame'>
Index: 520 entries, 0 to 519
Data columns (total 26 columns):
 #   Column                      Non-Null Count  Dtype         
---  ------                      --------------  -----         
 0   ContentID                   520 non-null    object        
 1   Title                       520 non-null    object        
 2   Type                        520 non-null    object        
 3   Genre                       520 non-null    object        
 4   Language                    520 non-null    object        
 5   Country                     520 non-null    object        
 6   Director                    516 non-null    object        
 7   CastCount                   520 non-null    int64         
 8   ReleaseYear                 520 non-null    int64         
 9   AddedDate                   517 non-null    datetime64[ns]
 10  RuntimeMinutes              519 non-null    float64       
 11  EpisodeCount                520 non-null    int64         
 12 